openhands/memory/condenser.py

```diff
@@ -1,24 +1,412 @@
+            return self.condense(state.history)
+
+    @classmethod
+    def from_config(cls, config: CondenserConfig) -> Condenser:
+        """Create a condenser from a configuration object.
+
+        Args:
+            config (CondenserConfig): Configuration for the condenser.
+
+        Returns:
+            Condenser: A condenser instance.
 
         Raises:
-        - Exception: the same exception as it got from the llm or processing the response
+            ValueError: If the condenser type is not recognized.
+        """
+        match config:
+            case NoOpCondenserConfig():
+                return NoOpCondenser()
+
+            case ObservationMaskingCondenserConfig():
+                return ObservationMaskingCondenser(
+                    **config.model_dump(exclude=['type'])
+                )
+
+            case RecentEventsCondenserConfig():
+                return RecentEventsCondenser(**config.model_dump(exclude=['type']))
+
+            case LLMSummarizingCondenserConfig(llm_config=llm_config):
+                return LLMSummarizingCondenser(llm=LLM(config=llm_config))
+
+            case AmortizedForgettingCondenserConfig():
+                return AmortizedForgettingCondenser(
+                    **config.model_dump(exclude=['type'])
+                )
+
+            case LLMAttentionCondenserConfig(llm_config=llm_config):
+                return LLMAttentionCondenser(
+                    llm=LLM(config=llm_config),
+                    **config.model_dump(exclude=['type', 'llm_config']),
+                )
+
+            case _:
+                raise ValueError(f'Unknown condenser config: {config}')
+
+
+class RollingCondenser(Condenser, ABC):
+    """Base class for a specialized condenser strategy that applies condensation to a rolling history.
+
+    The rolling history is computed by appending new events to the most recent condensation. For example, the sequence of calls::
+
+        assert state.history == [event1, event2, event3]
+        condensation = condenser.condensed_history(state)
+
+        # ...new events are added to the state...
+
+        assert state.history == [event1, event2, event3, event4, event5]
+        condenser.condensed_history(state)
+
+    will result in second call to `condensed_history` passing `condensation + [event4, event5]` to the `condense` method.
+    """
+
+    def __init__(self) -> None:
+        self._condensation: list[Event] = []
+        self._last_history_length: int = 0
+
+        super().__init__()
+
+    @override
+    def condensed_history(self, state: State) -> list[Event]:
+        new_events = state.history[self._last_history_length :]
+
+        with self.metadata_batch(state):
+            results = self.condense(self._condensation + new_events)
+
+        self._condensation = results
+        self._last_history_length = len(state.history)
+
+        return results
+
+
+@dataclass
+class CondensationObservation(Observation):
```

Comment:
In events/observation/agent, IIRC, is where the events of this kind are: not created exactly by the environment (e.g. not running in docker or stuff), but refer to things specific to the agent (-ish). For example, the old AgentSummarizeAction is in the events/action/ counterpart of that. Maybe we can move it?